src/main/java/org/broadinstitute/hellbender/tools/walkers/groundtruth/AddFlowBaseQuality.java

```diff
@@ -0,0 +1,335 @@
+import org.broadinstitute.hellbender.tools.FlowBasedArgumentCollection;
+import org.broadinstitute.hellbender.utils.read.FlowBasedRead;
+import org.broadinstitute.hellbender.utils.read.FlowBasedReadUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.SAMFileGATKReadWriter;
+import picard.cmdline.programgroups.ReadDataManipulationProgramGroup;
+
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Convert quality string that reports probability of indel errors (in flow chemistry) to the quality string that approximates probability of mismatch error.
+ *
+ * <p>
+ * </p>
+ *
+ * <p>
+ * The reference is strictly required when handling CRAM files.
+ * </p>
+ *
+ * <h3> Input </h3>
+ * <ul>
+ *     <li> Coordinate-sorted and indexed SAM/BAM/CRAM </li>
+ * </ul>
+ *
+ * <h3> Output </h3>
+ * <ul>
+ *     <li> Coordinate-sorted and indexed SAM/BAM/CRAM </li>
+ * </ul>
+ *
+ * {@GATK.walkertype ReadWalker}
+ */
+@CommandLineProgramProperties(
+        summary = "Prints reads from the input SAM/BAM/CRAM file to the SAM/BAM/CRAM file while adding a base quality attribute.",
+        oneLineSummary = "Add base quality attribute to reads in in the SAM/BAM/CRAM file",
+        programGroup = FlowBasedProgramGroup.class
+)
+@ExperimentalFeature
+@WorkflowProperties
+public final class AddFlowBaseQuality extends ReadWalker {
+
+    public static final String MINIMAL_ERROR_RATE_LONG_NAME = "minimal-error-rate";
+    public static final String MAXIMAL_QUALITY_SCORE_LONG_NAME = "maximal-quality-score";
+    public static final String REPLACE_QUALITY_MODE_LONG_NAME = "replace-quality-mode";
+    public static final String BASE_QUALITY_ATTRIBUTE_NAME = "BQ";
+    public static final String OLD_QUALITY_ATTRIBUTE_NAME = "OQ";
+    public static final char PHRED_ASCII_BASE = '!';
+
+    public static final int ERROR_PROB_BAND_1LESS = 0;
+    public static final int ERROR_PROB_BAND_KEY = 1;
+    public static final int ERROR_PROB_BAND_1MORE = 2;
+    public static final int ERROR_PROB_BANDS = 3;
+
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            doc="Write output to this file")
+    @WorkflowOutput(optionalCompanions={StandardArgumentDefinitions.OUTPUT_INDEX_COMPANION})
+    public GATKPath output;
+    private SAMFileGATKReadWriter outputWriter;
+
+    @Argument(fullName = MINIMAL_ERROR_RATE_LONG_NAME, doc = "lower floor for flow error rate values")
+    public double minErrorRate = 1e-3;
+
+    @Argument(fullName = MAXIMAL_QUALITY_SCORE_LONG_NAME, doc = "clip quality score to the given value (phred)")
+    public int maxQualityScore = 126;
+
+    @Argument(fullName = REPLACE_QUALITY_MODE_LONG_NAME, doc = "replace existing base qualities while saving previous qualities to OQ (when true) or simply write to BQ (when false) ")
```

Comment:
I'm not sure I understand the use case of writing to BQ without changing the actual base qualities. Both of these options make the file larger and contain the same information (since you continue to store the old base qualities either in the QUAL field or in the OQ tag and you store the new qualities either in the BQ tag or the QUAL field). Is there a reason to not replace the QUAL field and only keep the new qualities in the BQ tag? 

Additionally the BQ tag is reserved in the spec for "Offset to base alignment quality (BAQ)" so might not be the best choice to put these new base qualities.

Reply:
As I understand the two modes, they do not involve storing the same information twice.

In the 'replace' mode, the QUAL field is replaced with qualities computed from the flow probabilities .The old QUAL field is saved in OQ.

In the 'non replace mode', the QUAL field is preserved while a new quality string, computed from the flow probabilities, is saved in BQ.

It is assumed that computed probabilities will be different than the original QUAL - giving rise to this tool.

As to the BQ being reserved, I have replaced with XQ, which is the user defined space.